lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,872 @@
+//===-- lib/lower/ast-builder.cc ------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/lower/ASTBuilder.h"
+#include "flang/parser/parse-tree-visitor.h"
+#include <cassert>
+#include <utility>
+
+/// Build a light-weight AST to help with lowering to FIR.  The AST will
+/// capture pointers back into the parse tree, so the parse tree data structure
+/// may <em>not</em> be changed between the construction of the AST and all of
+/// its uses.
+///
+/// The AST captures a structured view of the program.  The program is a list of
+/// units.  Function like units will contain lists of evaluations.  Evaluations
+/// are either statements or constructs, where a construct contains a list of
+/// evaluations.  The resulting AST structure can then be used to create FIR.
+
+namespace Br = Fortran::lower;
+namespace Co = Fortran::common;
+namespace L = llvm;
+namespace Pa = Fortran::parser;
```

Comment:
Code is easier to understand if the same names are used everywhere. Is `Pa` really that much easier to type than `parser`?

Reply:
This was commented upon during a previous review of this (or similar) code.

Are the reviewers of this pull request expected to perform a full review, or should our comments be limited to new items?

Reply:
We're going to use short name aliases to avoid proliferation of using namespaces and typing long namespace identifiers everywhere.

Reply:
_Originally posted by @DavidTruby in https://github.com/flang-compiler/f18/diffs_

I don't think these aid readability in this file.

Reply:
Just to elaborate on what I meant:
I think that when diving into the middle of these files (imagine a bug-fix PR or similar review) it's unclear when immediately reading what the shortened namespaces actually mean. 
I think e.g. `namespace parser = Fortran::parser;` is a lot clearer and not a significantly longer namespace identifier.

Reply:
This follows the same principle as to why England, London, and Epping Upland don't appear on a map in the same sized font. Let me explain further.

If one applied the "type everything all the way out" rule universally, we would write something like
```
   Physics::ideal_gases::pressure *
      Physics::ideal_gases::volume =
   Physics::mass::number_of_molecules *
      Physics::constant::Avogadros_number *
      Physics::ideal_gases::temperature
```

Of course, one could argue there are intermediate steps that are just as good, such as
```
   using Physics;
   ideal_gases::pressure * ideal_gases::volume =
      mass::number_of_molecules * constant::Avogadros_number *
      ideal_gases::temperature
```
and so forth. (This leads to a trap. For example, names might be reused many times. Use of the identifier `Expression` might be defined in `Fortran::parser`, `Fortran::evaluate`, `Fortran::semantics`, `mlir`, and `llvm`, this making comprehension of a bare `Expression` more difficult.)

Now, if we look in a copy of "Intro to Physics", we will see the above formula expressed as
```
   Ideal Gas Law:
      P V = n R T
```

Why would the author/editor/publisher present this in such a terse form when we all should assume that long names are better or, at least, just as readable?

The last form, as presented to someone new to the subject, increases understanding, comprehension, and memory retention. In short [pun intended], it's brevity within a given context makes it far more "readable." (There is science behind this. Shapes, pattern matching, relative salience, visual scanning, short-term memory, short-term to long-term memory transfer. Call it how the human brain works.) We know that students will long remember the formula "PV=nRT" decades after their one (and perhaps only) Physics class and longer than the remember the value of Avogadro's number or Boltzmann's constant.

Reply:
Why not use `P` then rather than `Pa` for `parser`, if this actually is a good idea?

Reply:
At an earlier time, I had a collision where two namespaces trivially mapped to the same alias letter. We could shorten things to one letter whenever possible, though the current convention is used ubiquitously throughout the project, most of which is _not_ this PR.

Reply:
I would prefer spelled-out short namespace names like `parser`.  They don't have to be decoded by the reader.

Reply:
Abbreviations have been removed.